lib/gcov/ukgcov.c

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: MIT */
+/* Copyright © 2005-2020 Rich Felker, et al.
+ * Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the MIT License (the "License", see COPYING.md).
+ * You may not use this file except in compliance with the License.
+ */
+/*
+ * The code in this file was derived from gcc-13-onlinedocs:
+ * Source: https://gcc.gnu.org/onlinedocs/gcc/Freestanding-Environments.html
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include "ukgcov.h"
+
+#ifdef GCOV_OPT_OUTPUT_BINARY_MEMORY
+static unsigned char __attribute__ ((used)) *gcov_output_buffer;
+static __attribute__ ((used)) size_t gcov_output_buffer_pos;
+static size_t gcov_output_buffer_size = INITIAL_BUFFER_SIZE;
+#endif
+
+/* This function shall produce a reliable in order byte stream to transfer the
+   gcov information from the target to the host system.  */
+static void dump(const void *d, unsigned n, void *arg)
+{
+#ifdef GCOV_OPT_OUTPUT_BINARY_MEMORY
```

Comment:
It would be cleaner to move the implementation of each output type to its own static function, and call it from here. Similarly for the dump function.